File: servlets/src/java/edu/rice/rubis/servlets/StoreBid.java
/*
 * Copyright (C) 2002-2009  OW2 Consortium
 *
 * This file is part of dcsj-rubis (below referred to as "this program").
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package edu.rice.rubis.servlets;

import java.io.IOException;
import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;

import javax.servlet.ServletException;
import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;

/**
 * This servlet records a bid in the database and display
 * the result of the transaction.
 * It must be called this way :
 * <pre>
 * http://..../StoreBid?itemId=aa&userId=bb&minBid=cc&maxQty=dd&bid=ee&maxBid=ff&qty=gg 
 *   where: aa is the item id 
 *          bb is the user id
 *          cc is the minimum acceptable bid for this item
 *          dd is the maximum quantity available for this item
 *          ee is the user bid
 *          ff is the maximum bid the user wants
 *          gg is the quantity asked by the user
 * </pre>
 *
 * @author <a href="mailto:cecchet@rice.edu">Emmanuel Cecchet</a>
 * @author <a href="mailto:julie.marguerite@inrialpes.fr">Julie Marguerite</a>
 * @author <a href="mailto:marco.guazzone@gmail.com">Marco Guazzone</a>
 */

public class StoreBid extends RubisHttpServlet
{
	@Override
	protected void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException
	{
		this.doPost(request, response);
	}

	@Override
	protected void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException
	{
		int userId; // item id
		int itemId; // user id
		float minBid; // minimum acceptable bid for this item
		float bid; // user bid
		float maxBid; // maximum bid the user wants
		int maxQty; // maximum quantity available for this item
		int qty; // quantity asked by the user
		ServletPrinter sp = null;
		PreparedStatement stmt = null;
		Connection conn = null;

		sp = new ServletPrinter(response, "StoreBid");

		/* Get and check all parameters */

		String value = request.getParameter("userId");
		if ((value == null) || (value.equals("")))
		{
			this.printError("You must provide a user identifier!", sp);
			return;
		}
		userId = Integer.parseInt(value);

		value = request.getParameter("itemId");
		if ((value == null) || (value.equals("")))
		{
			this.printError("You must provide an item identifier!", sp);
			return;
		}
		itemId = Integer.parseInt(value);

		value = request.getParameter("minBid");
		if ((value == null) || (value.equals("")))
		{
			this.printError("You must provide a minimum bid!", sp);
			return;
		}
		minBid = Float.parseFloat(value);

		value = request.getParameter("bid");
		if ((value == null) || (value.equals("")))
		{
			this.printError("You must provide a bid!", sp);
			return;
		}
		bid = Float.parseFloat(value);

		value = request.getParameter("maxBid");
		if ((value == null) || (value.equals("")))
		{
			this.printError("You must provide a maximum bid!", sp);
			return;
		}
		maxBid = Float.parseFloat(value);

		value = request.getParameter("maxQty");
		if ((value == null) || (value.equals("")))
		{
			this.printError("You must provide a maximum quantity!", sp);
			return;
		}
		maxQty = Integer.parseInt(value);

		value = request.getParameter("qty");
		if ((value == null) || (value.equals("")))
		{
			this.printError("You must provide a quantity!", sp);
			return;
		}
		qty = Integer.parseInt(value);

		/* Check for invalid values */

		if (maxQty <= 0 || qty > maxQty)
		{
			this.printError("You cannot request " + qty + " items because only " + maxQty + " are proposed!", sp);
			return;
		}
		if (bid < minBid)
		{
			this.printError("Your bid of $" + bid + " is not acceptable because it is below the $" + minBid + " minimum bid!", sp);
			return;
		}
		if (maxBid < minBid)
		{
			this.printError("Your maximum bid of $" + maxBid + " is not acceptable because it is below the $" + minBid + " minimum bid!", sp);
			return;
		}
		if (maxBid < bid)
		{
			this.printError("Your maximum bid of $" + maxBid + " is not acceptable because it is below your current bid of $" + bid + "!", sp);
			return;
		}
		try
		{
			conn = this.getConnection();
			conn.setAutoCommit(false);
			String now = TimeManagement.currentDateToString();
			stmt = conn.prepareStatement( "INSERT INTO bids VALUES (NULL,?,?,?,?,?,?)");
			stmt.setInt(1, userId);
			stmt.setInt(2, itemId);
			stmt.setInt(3, qty);
			stmt.setFloat(4, bid);
			stmt.setFloat(5, maxBid);
			stmt.setString(6, now);
			stmt.executeUpdate();
			stmt.close();
			// update the number of bids and the max bid for the item
			PreparedStatement update = null;
			try
			{
				stmt = conn.prepareStatement("SELECT nb_of_bids, max_bid FROM items WHERE id=?");
				stmt.setInt(1, itemId);
				ResultSet rs = stmt.executeQuery();
				if (rs.first())
				{
					int nbOfBids = rs.getInt("nb_of_bids");
					nbOfBids++;
					float oldMaxBid = rs.getFloat("max_bid");
					if (bid > oldMaxBid)
					{
						oldMaxBid = bid;
						update =
						conn.prepareStatement("UPDATE items SET max_bid=?, nb_of_bids=? WHERE id=?");
						update.setFloat(1, maxBid);
						update.setInt(2, nbOfBids);
						update.setInt(3, itemId);
						update.executeUpdate();
						update.close();
					}
					else
					{
						update = conn.prepareStatement("UPDATE items SET nb_of_bids=? WHERE id=?");
						update.setInt(1, nbOfBids);
						update.setInt(2, itemId);
						update.executeUpdate();
						update.close();
					}
				}
				else
				{
					conn.rollback();
					this.printError("Couldn't find the item.", sp);
					this.closeConnection(stmt, conn);
					return;
				}
			}
			catch (Exception ex)
			{
				conn.rollback();
				this.printError("Failed to update nb of bids and max bid: " + ex, sp);
				if (update != null) 
				{
					update.close();
				}
				this.closeConnection(stmt, conn);
				return;
			}
			conn.commit();

			sp.printHTMLheader("RUBiS: Bidding result");
			sp.printHTML("<center><h2>Your bid has been successfully processed.</h2></center>\n");
			sp.printHTMLfooter();
		}
		catch (Exception e)
		{
			this.printError("Error while storing the bid: " + e, sp);
			try
			{
				conn.rollback();
			}
			catch (Exception se)
			{
				this.printError("Transaction rollback failed: " + e, sp);
			}
		}
		this.closeConnection(stmt, conn);
	}

	@Override
	protected int getPoolSize()
	{
		return Config.StoreBidPoolSize;
	}

	/**
	 * Close both statement and connection.
	 */
	private void closeConnection(PreparedStatement stmt, Connection conn)
	{
		try
		{
			if (stmt != null)
			{
				stmt.close(); // close statement
			}
			if (conn != null)
			{
				conn.setAutoCommit(true);
				this.releaseConnection(conn);
			}
		}
		catch (Exception ignore)
		{
		}
	}

	/**
	 * Display an error message.
	 * @param errorMsg the error message value
	 */
	private void printError(String errorMsg, ServletPrinter sp)
	{
		this.printError("Store Bid", errorMsg, sp);
	}
}

File: php/StoreBid.php
<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
  <body>
    <?php
    $scriptName = "StoreBid.php";
    include("PHPprinter.php");
    $startTime = getMicroTime();
    
	$userId = NULL;
	if (isset($_POST['userId']))
	{
    	$userId = $_POST['userId'];
	}
	else if (isset($_GET['userId']))
	{
    	$userId = $_GET['userId'];
	}
	else
	{
		printError($scriptName, $startTime, "StoreBid", "<h3>You must provide a user identifier!<br></h3>");
		exit();
	}
	$itemId = NULL;
	if (isset($_POST['itemId']))
	{
    	$itemId = $_POST['itemId'];
	}
	else if (isset($_GET['itemId']))
	{
    	$itemId = $_GET['itemId'];
	}
	else
	{
		printError($scriptName, $startTime, "StoreBid", "<h3>You must provide an item identifier !<br></h3>");
		exit();
	}
	$minBid = NULL;
	if (isset($_POST['minBid']))
	{
    	$minBid = $_POST['minBid'];
	}
	else if (isset($_GET['minBid']))
	{
    	$minBid = $_GET['minBid'];
	}
	else
	{
		printError($scriptName, $startTime, "StoreBid", "<h3>You must provide an item identifier !<br></h3>");
		exit();
	}
	$bid = NULL;
	if (isset($_POST['bid']))
	{
    	$bid = $_POST['bid'];
	}
	else if (isset($_GET['bid']))
	{
    	$bid = $_GET['bid'];
	}
	else
	{
		printError($scriptName, $startTime, "StoreBid", "<h3>You must provide a minimum bid !<br></h3>");
		exit();
	}
	$maxBid = NULL;
	if (isset($_POST['maxBid']))
	{
    	$maxBid = $_POST['maxBid'];
	}
	else if (isset($_GET['maxBid']))
	{
    	$maxBid = $_GET['maxBid'];
	}
	else
	{
		printError($scriptName, $startTime, "StoreBid", "<h3>You must provide a maximum bid !<br></h3>");
		exit();
	}
	$maxQty = NULL;
	if (isset($_POST['maxQty']))
	{
    	$maxQty = $_POST['maxQty'];
	}
	else if (isset($_GET['maxQty']))
	{
    	$maxQty = $_GET['maxQty'];
	}
	else
	{
		printError($scriptName, $startTime, "StoreBid", "<h3>You must provide a maximum quantity !<br></h3>");
		exit();
	}
	$qty = NULL;
	if (isset($_POST['qty']))
	{
    	$qty = $_POST['qty'];
	}
	else if (isset($_GET['qty']))
	{
    	$qty = $_GET['qty'];
	}
	else
	{
		printError($scriptName, $startTime, "StoreBid", "<h3>You must provide a quantity !<br></h3>");
		exit();
	}

    /* Check for invalid values */

    if ($qty > $maxQty)
    {
      printError("<h3>You cannot request $qty items because only $maxQty are proposed !<br></h3>");
      return ;
    }      
    if ($bid < $minBid)
    {
      printError("<h3>Your bid of \$$bid is not acceptable because it is below the \$$minBid minimum bid !<br></h3>");
      return ;
    }      
    if ($maxBid < $minBid)
    {
      printError("<h3>Your maximum bid of \$$maxBid is not acceptable because it is below the \$$minBid minimum bid !<br></h3>");
      return ;
    }      
    if ($maxBid < $bid)
    {
      printError("<h3>Your maximum bid of \$$maxBid is not acceptable because it is below your current bid of \$$bid !<br></h3>");
      return ;
    }      

    getDatabaseLink($link);
    begin($link);

    // Add bid to database and update values in item
    $now = date("Y:m:d H:i:s");
//    $result = mysql_query("LOCK TABLES bids WRITE, items WRITE", $link);
//	if (!$result)
//	{
//		error_log("[".__FILE__."] Failed to acquire locks on items and bids tables: " . mysql_error($link));
//		die("ERROR: Failed to acquire locks on items and bids tables: " . mysql_error($link));
//	}
    $result = mysql_query("SELECT max_bid FROM items WHERE id=$itemId", $link);
	if (!$result)
	{
		error_log("[".__FILE__."] Query 'SELECT max_bid FROM items WHERE id=$itemId' failed: " . mysql_error($link) . ". DEADLOCK!!");
		die("ERROR: Failed to query the number of bids for item '$itemId' in database: " . mysql_error($link) . ". DEADLOCK!!");
	}
    $row = mysql_fetch_array($result);
    if ($maxBid > $row["max_bid"])
	{
      $result2 = mysql_query("UPDATE items SET max_bid=$maxBid WHERE id=$itemId", $link);
	  if (!$result2)
	  {
		error_log("[".__FILE__."] Failed to update maximum bid in database 'UPDATE items SET max_bid=$maxBid WHERE id=$itemId': " . mysql_error($link) . ". DEADLOCK!!");
		die("ERROR: Failed to update maximum bid '$maxBid' for item '$itemId' in database: " . mysql_error($link) . ". DEADLOCK!!");
	  }
	}

    $result = mysql_query("INSERT INTO bids VALUES (NULL, $userId, $itemId, $qty, $bid, $maxBid, '$now')", $link);
	if (!$result)
	{
		error_log("[".__FILE__."] Failed to insert new bid in database 'INSERT INTO bids VALUES (NULL, $userId, $itemId, $qty, $bid, $maxBid, '$now')': " . mysql_error($link) . ". DEADLOCK!!");
		die("ERROR: Failed to insert new bid for user '$userId' and item '$itemId' in database: " . mysql_error($link) . ". DEADLOCK!!");
	}
    $result = mysql_query("UPDATE items SET nb_of_bids=nb_of_bids+1 WHERE id=$itemId", $link);
	if (!$result)
	{
		error_log("[".__FILE__."] Failed to update number of bids in database 'UPDATE items SET nb_of_bids=nb_of_bids+1 WHERE id=$itemId': " . mysql_error($link) . ". DEADLOCK!!");
		die("ERROR: Failed to update number of bids for item '$itemId' in database: " . mysql_error($link) . ". DEADLOCK!!");
	}
//    $result = mysql_query("UNLOCK TABLES", $link);
//	if (!$result)
//	{
//		error_log("[".__FILE__."] Failed to unlock items and bids tables: " . mysql_error($link));
//		die("ERROR: Failed to unlock items and bids tables: " . mysql_error($link));
//	}
    commit($link);

    printHTMLheader("RUBiS: Bidding result");
    print("<center><h2>Your bid has been successfully processed.</h2></center>\n");
    
    mysql_close($link);
    
    printHTMLfooter($scriptName, $startTime);
    ?>
  </body>
</html>
